### PR TITLE
mfa: audit events for adding/removing devices

### PR DESCRIPTION
**Pull request by awly**
Also add `With2FA` field to `UserLogin` when an MFA device is used
during login.

### CODE REVIEW

File: api/types/events/events.proto [hunk 1]
```diff
@@ -1173,6 +1176,44 @@ message DatabaseSessionEnd {
...
+
+// MFADeviceDelete is emitted when a user deletes an MFA device.
+message MFADeviceDelete {
+    // Metadata is a common event metadata.
+    Metadata Metadata = 1
+        [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+    // User is a common user event metadata.
+    UserMetadata User = 2
+        [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+    // Device is the new MFA device deleted by the user.
```
Comment (Joerger) [inline api/types/events/events.proto:1213]: nit: <s>new</s>

File: lib/auth/auth.go [hunk 2]
```diff
@@ -57,12 +57,13 @@ import (
...
 	"github.com/coreos/go-oidc/oauth2"
 	"github.com/coreos/go-oidc/oidc"
-	"github.com/gravitational/trace"
 	"github.com/jonboulle/clockwork"
 	"github.com/pborman/uuid"
 	"github.com/prometheus/client_golang/prometheus"
 	saml2 "github.com/russellhaering/gosaml2"
 	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/trace"
```
Comment (Joerger) [inline lib/auth/auth.go:66]: nit: Why'd you separate this from the import group? I thought `trace` was supposed to be separated from the main teleport import group.

File: lib/auth/grpcserver.go [hunk 3]
```diff
@@ -38,13 +46,6 @@ import (
 	"github.com/gravitational/teleport/lib/utils"
 	"github.com/gravitational/trace"
```
Comment (Joerger) [inline lib/auth/grpcserver.go:47]: nit: Same question about `trace`

File: lib/auth/helpers.go [hunk 4]
```diff
@@ -40,8 +40,9 @@ import (
 	"github.com/gravitational/teleport/lib/session"
 	"github.com/gravitational/teleport/lib/utils"
 
-	"github.com/gravitational/trace"
 	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/trace"
```
Comment (Joerger) [inline lib/auth/helpers.go:45]: nit: same question as above.

**Review by Joerger — APPROVED** [review]

File: lib/auth/password.go [hunk 5]
```diff
@@ -190,59 +196,58 @@ func (s *Server) CheckPasswordWOToken(user string, password []byte) error {
 }
 
 // CheckPassword checks the password and OTP token. Called by tsh or lib/web/*.
-func (s *Server) CheckPassword(user string, password []byte, otpToken string) error {
+func (s *Server) CheckPassword(user string, password []byte, otpToken string) (*types.MFADevice, error) {
```
Comment (r0mant) [inline lib/auth/password.go:199]: Hmm, this makes the function signature a little strange IMO. I.e. it seems a little "random" that CheckPassword returns MFA device, though I understand why it's needed.

As a small improvement suggestion, what if we made this function return `CheckPasswordResult` struct that MFADevice would be one of the fields of (and the only field, for now at least)? Then I think the intent will be a bit clearer, and you wouldn't also need to return `nil, nil` from this function (as it is now).

File: lib/auth/password.go [hunk 6]
```diff
@@ -190,59 +196,58 @@ func (s *Server) CheckPasswordWOToken(user string, password []byte) error {
...
 	err := s.CheckPasswordWOToken(user, password)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	err = s.checkOTP(user, otpToken)
-	return trace.Wrap(err)
+	return s.checkOTP(user, otpToken)
```
Comment (xacrimon) [inline lib/auth/password.go:205]: Shouldn't we wrap the error here to get trace context?

**Review by xacrimon — APPROVED** [review]

File: api/types/events/events.proto [hunk 7]
```diff
@@ -519,6 +519,9 @@ message UserLogin {
     // IdentityAttributes is a map of user attributes received from identity provider
     google.protobuf.Struct IdentityAttributes = 5
         [ (gogoproto.jsontag) = "attributes,omitempty", (gogoproto.casttype) = "Struct" ];
+
+    // With2FA is the UUID of an MFA device used during the login.
+    string With2FA = 6 [ (gogoproto.jsontag) = "with_2fa,omitempty" ];
```
[the remark this answers is not in the document]
Comment (awly) [reply api/types/events/events.proto]: Makes sense, added

File: lib/auth/auth.go (hunk 2 above)
Comment (awly) [reply lib/auth/auth.go]: I mis-configured `gopls` import grouping in my editor, reverted :)

File: lib/auth/password.go (hunk 5 above)
Comment (awly) [reply lib/auth/password.go]: Done

**Comment by russjones** [comment]
@r0mant Do you mind re-reviewing? Since this is slated for 6.1 if you don't have time now it can wait till after 6.0 is out.

**Comment by awly** [comment]
@r0mant please review

File: lib/auth/methods.go [hunk 8]
```diff
@@ -110,41 +114,50 @@ func (s *Server) AuthenticateUser(req AuthenticateUserRequest) error {
...
+			return err
 		})
 		if err != nil {
 			// provide obscure message on purpose, while logging the real
 			// error server side
 			log.Debugf("Failed to authenticate: %v.", err)
-			return trace.AccessDenied("invalid U2F response")
+			return nil, trace.AccessDenied("invalid U2F response")
 		}
-		return nil
```
Comment (r0mant) [inline lib/auth/methods.go:136]: Nit: I would add `if mfaDev == nil { return nil, trace.XXX }` check here and below, just in case.

**Review by r0mant — APPROVED** [review]

**Pull request merged**